habitat_baselines/common/obs_transformers.py

```diff
@@ -217,113 +218,496 @@ def from_config(cls, config: Config):
+        # Obtain points on unit sphere
+        world_pts, not_assigned_mask = self.output_model.unprojection()
         # Generate grid
         grids = []
-        for rot in rotations:
-            R = torch.from_numpy(rot.T).float()
-            rotate_on_sphere = torch.matmul(
-                xyz_on_sphere.view((-1, 3)), R
-            ).view(equ_h, equ_w, 3)
-
-            # Project points on z=1 plane
-            grid = rotate_on_sphere / torch.abs(rotate_on_sphere[..., 2:3])
-            mask = torch.abs(grid).max(-1)[0] <= 1  # -1 <= grid.xy <= 1
-            mask *= grid[..., 2] == 1
-            grid[
-                ~mask
-            ] = 2  # values bigger than one will be ignored by grid_sample
-            grid_xy = -grid[..., :2].unsqueeze(0)
-            grids.append(grid_xy)
-        grids = torch.cat(grids, dim=0)
+        for input_model in self.input_models:
+            grid, input_mask = input_model.projection(world_pts)
+            # Make sure each point is only assigned to single input
+            input_mask *= not_assigned_mask
+            # Values bigger than one will be ignored by grid_sample
+            grid[~input_mask] = 2
+            # Update not_assigned_mask
+            not_assigned_mask *= ~input_mask
+            grids.append(grid)
+        grids = torch.stack(grids, dim=0)
         return grids
 
-    def _to_equirec(self, batch: torch.Tensor) -> torch.Tensor:
-        """Takes a batch of cubemaps stacked in proper order and converts thems to equirects, reduces batch size by 6"""
+    def _convert(self, batch: torch.Tensor) -> torch.Tensor:
+        """Takes a batch of images stacked in proper order and converts thems, reduces batch size by input_len"""
         batch_size, ch, _H, _W = batch.shape
-        if batch_size == 0 or batch_size % 6 != 0:
-            raise ValueError("Batch size should be 6x")
+        input_len = len(self.input_models)
+        if batch_size == 0 or batch_size % input_len != 0:
+            raise ValueError(f"Batch size should be {input_len}x")
         output = torch.nn.functional.grid_sample(
             batch,
             self._grids_cache,
             align_corners=True,
             padding_mode="zeros",
         )
         output = output.view(
-            batch_size // 6, 6, ch, self.equ_h, self.equ_w
+            batch_size // input_len,
+            input_len,
+            ch,
+            self.output_model.img_h,
+            self.output_model.img_w,
         ).sum(dim=1)
-        return output  # batch_size // 6, ch, self.equ_h, self.equ_w
+        return output  # batch_size // input_len, ch, output_model.img_h, output_model.img_w
 
-    # Convert input cubic tensor to output equirectangular image
-    def to_equirec_tensor(self, batch: torch.Tensor) -> torch.Tensor:
-        batch_size = batch.size()[0]
+    def to_converted_tensor(self, batch: torch.Tensor) -> torch.Tensor:
+        batch_size, ch = batch.size()[:2]
```

Comment:
This assumes channels first, which I suppose is for this, but there should be a comment in this regard.

Reply:
I added the order `NCHW` in a comment.